Blueprint: describe directory and file customizations

thozza opened the pull request:
Explain how directory and file BP customizations can be used and what are the restrictions that apply when using it.

thozza commented:
This is still blocked on the feature landing in osbuild-composer.

achilleas-k commented on osbuild-composer/src/blueprint-reference/blueprint-reference.md:
```diff
@@ -227,6 +227,72 @@ The service names are systemd service units. You may specify any systemd unit fi
+When using the custom files and directories customization, the following rules apply:
+
+- The path must be an absolute path and must be under `/etc`.
+- There may be no duplicate paths of the same type (file or directory) nor may there be a file and directory with the same path.
```
This is a little confusing.  I think mentioning the type ("... of the same type ... nor file and directory") makes it unclear because essentially what we want to say is "you can't have the same path more than once" and that we could file and directory paths together.

thozza replied:
I'll try to reword it a bit... 🤔

thozza replied:
Thanks for the suggestion, I'll go with it...

achilleas-k commented on osbuild-composer/src/blueprint-reference/blueprint-reference.md:
```diff
@@ -227,6 +227,72 @@ The service names are systemd service units. You may specify any systemd unit fi
+
+#### Custom files
+
+Custom files can be created by specifying items in the `customizations.files` list. The customization can be used to create new files or to replace existing ones, if not restricted by the policy. Note that if the target path is an existing symlink to another file, it will be replaced by the custom file.
```
```suggestion
Custom files can be created by specifying items in the `customizations.files` list. The customization can be used to create new files or to replace existing ones, if not restricted by the policy. Note that if the target path is an existing symlink to another file, the symlink will be replaced by the custom file.
```
I think **it** here is ambiguous: it may refer to the "existing symlink" or the "another file".

achilleas-k commented on osbuild-composer/src/blueprint-reference/blueprint-reference.md:
```diff
@@ -227,6 +227,72 @@ The service names are systemd service units. You may specify any systemd unit fi
+- `/etc/passwd`
+- `/etc/group`
+
+Using the `files` customization comes with a high chance of breaking the image. It is recommended to use it only if you know what you are doing. Although the `files` customization can be used to configure parts of the OS, that can be done by other blueprint customizations, this use is discouraged. If possible, users should always default to using the specialized blueprint customizations. Note tha ...
```
```suggestion
Using the `files` customization comes with a high chance of creating an image that doesn't boot. **Use this feature only if you know what you are doing**. Although the `files` customization can be used to configure parts of the OS which can also be configured by other blueprint customizations, this use is discouraged. If possible, users should always default to using the specialized blueprint customizations. Note that if you combine the files customizations with other customizations, the other customizations may not work as expected or may be overridden by the files customizations.
```

Some minor rephrasing, if you don't mind :)

achilleas-k reviewed (commented):
Thanks.  Good write-up.  Some suggestions below.

ondrejbudai commented on osbuild-composer/src/blueprint-reference/blueprint-reference.md:
```diff
@@ -227,6 +227,72 @@ The service names are systemd service units. You may specify any systemd unit fi
+
+- The path must be an absolute path and must be under `/etc`.
+- There may be no duplicate paths of the same type (file or directory) nor may there be a file and directory with the same path.
+- No custom file can be a parent of another custom file or directory.
```
I wonder if this line isn't just noise and if it should be removed. Files just generally cannot be parents of anything.

thozza replied:
I do not have any preference here. Happy to remove it. I basically described all explicit checks that we do in the code.

ondrejbudai reviewed (commented):
Looks good! I wonder whether we should also include a note about stage ordering - specifically that the service stage is run after custom files are added so one can use this customization to create custom services.

elianerpereira commented on osbuild-composer/src/blueprint-reference/blueprint-reference.md:
```diff
@@ -227,6 +227,72 @@ The service names are systemd service units. You may specify any systemd unit fi
+
+#### Custom files
+
+Custom files can be created by specifying items in the `customizations.files` list. The customization can be used to create new files or to replace existing ones, if not restricted by the policy. Note that if the target path is an existing symlink to another file, the symlink will be replaced by the custom file.
```
As a suggestion, you could use the direct voice: "You can create custom files by specifying items in the `customizations.files` list."

elianerpereira reviewed (commented):
Hi, I finished my review and provided a few suggestions to you. Please, feel free to adopt them or not :-) The content is great content and you added very good examples. Thank you for that!

thozza commented on osbuild-composer/src/blueprint-reference/blueprint-reference.md:
```diff
@@ -227,6 +227,72 @@ The service names are systemd service units. You may specify any systemd unit fi
+
+### Custom files and directories
+
+Blueprint customizations can be used to create custom files and directories in the image. This ability is currently restricted to only `/etc`.
```
Thanks for the great suggestion.

thozza commented on osbuild-composer/src/blueprint-reference/blueprint-reference.md:
```diff
@@ -227,6 +227,72 @@ The service names are systemd service units. You may specify any systemd unit fi
+
+#### Custom directories
+
+Custom directories can be created by specifying items in the `customizations.directories` list. The customization can be used to create new directories or to modify existing ones, if not restricted by the policy.
```
Thanks for the suggestion. As for the policy, I meant what is on line 235. In addition, we were considering restricting paths which can not be modified with this customization, but in the end we do not allow modifying any existing directories. So I'll just reflect that and won't mention any policy.

thozza commented:
Thanks to all for review. I addressed all review comments.

While reading through the documentation again, it got me thinking whether I should:

1. Document for which image types this is not supported? Those would be {iot,edge}-{raw-image,installer,simplified-installer}. However we don't do it for any other customizations which are not supported for these image types. So I'm not sure if it makes sense to add such note for these two customizations or whether we want to document this per image type at some point.
2. Describe the implications for ostree images? E.g. that the custom files and directories will disappear if the upgrade commits built with image builder do not contain the same files and directories customizations. But this is not exclusive to these customizations, since a different or dropped customizations would behave the same way. So this may be again something to document (if at all) for specific image types.

ondrejbudai commented:
> Thanks to all for review. I addressed all review comments.
> 
> While reading through the documentation again, it got me thinking whether I should:
> 
>     1. Document for which image types this is not supported? Those would be {iot,edge}-{raw-image,installer,simplified-installer}. However we don't do it for any other customizations which are not supported for these image types. So I'm not sure if it makes sense to add such note for these two customizations or whether we want to document this per image type at some point.

I would go for it. Better be explicit than have surprised users.

> 
>     2. Describe the implications for ostree images? E.g. that the custom files and directories will disappear if the upgrade commits built with image builder do not contain the same files and directories customizations. But this is not exclusive to these customizations, since a different or dropped customizations would behave the same way. So this may be again something to document (if at all) for specific image types.

Yeah, I think this should go into some kind of generic documentation for ostree-based images. Not sure where this should live to be honest. rpm-ostree? Fedora IoT? Our docs?

achilleas-k commented on osbuild-composer/src/blueprint-reference/blueprint-reference.md:
```diff
@@ -227,6 +227,70 @@ The service names are systemd service units. You may specify any systemd unit fi
+When using the custom files and directories customization, the following rules apply:
+
+- The path must be an absolute path and must be under `/etc`.
+- There might be no duplicate paths of the same directory.
+- There might be no duplicated paths of the same file.
```
Why did this change to "might" instead of "must"?  "can" also works (There can be no duplicate...), but "might" doesn't sound right.

Also, one says duplicate and the other duplicated.  I think either works, but should be the same.

thozza replied:
I changed it to `must`, since that was used in the first list item.

achilleas-k commented on osbuild-composer/src/blueprint-reference/blueprint-reference.md:
```diff
@@ -227,6 +227,70 @@ The service names are systemd service units. You may specify any systemd unit fi
+
+You can create custom files by specifying items in the `customizations.files` list. You can use the customization to create new files or to replace existing ones, if not restricted by the policy specified below. If the target path is an existing symlink to another file, the symlink will be replaced by the custom file.
+
+Please note that the parent directory of a specified file must exist. If it does not exist, the image build will fail. One can assure that the parent directory exists by specifying it in the `customizations.directories` list.
```
```suggestion
Please note that the parent directory of a specified file must exist. If it does not exist, the image build will fail. One can ensure that the parent directory exists by specifying it in the `customizations.directories` list.
```

thozza commented:
> > Thanks to all for review. I addressed all review comments.
> > While reading through the documentation again, it got me thinking whether I should:
> > ```
> > 1. Document for which image types this is not supported? Those would be {iot,edge}-{raw-image,installer,simplified-installer}. However we don't do it for any other customizations which are not supported for these image types. So I'm not sure if it makes sense to add such note for these two customizations or whether we want to document this per image type at some point.
> > ```
> 
> I would go for it. Better be explicit than have surprised users.

I added a note.

> 
> > ```
> > 2. Describe the implications for ostree images? E.g. that the custom files and directories will disappear if the upgrade commits built with image builder do not contain the same files and directories customizations. But this is not exclusive to these customizations, since a different or dropped customizations would behave the same way. So this may be again something to document (if at all) for specific image types.
> > ```
> 
> Yeah, I think this should go into some kind of generic documentation for ostree-based images. Not sure where this should live to be honest. rpm-ostree? Fedora IoT? Our docs?

Our docs in composer (?) describing supported image types would probably make sense. I'll leave this for my (or someone else's) future self 😄

ondrejbudai commented on osbuild-composer/src/blueprint-reference/blueprint-reference.md:
```diff
@@ -227,6 +227,72 @@ The service names are systemd service units. You may specify any systemd unit fi
+
+### Custom files and directories
+
+You can use Blueprint customizations to create custom files and directories in the image. These customizations are currently restricted only to the `/etc` directory.
```
It seems like we don't capitalize blueprints anywhere else, so I would in favor of using *blueprint* here.

thozza replied:
sure...

ondrejbudai reviewed (commented):
One tiny nitpick, otherwise this looks great!

ondrejbudai reviewed (approved):
Perfect, thanks!

achilleas-k commented:
Awesome.  Thank you!!
